cli/tests/websocketstream_test.ts

```diff
@@ -0,0 +1,123 @@
+// Copyright 2018-2021 the Deno authors. All rights reserved. MIT license.
+// deno-lint-ignore-file no-undef
+import {
+  assertEquals,
+  assertThrows,
+  assertThrowsAsync,
+} from "../../test_util/std/testing/asserts.ts";
+
+Deno.test("invalid scheme", () => {
+  assertThrows(() => new WebSocketStream("foo://localhost:4242"));
+});
+
+Deno.test("fragment", () => {
+  assertThrows(() => new WebSocketStream("ws://localhost:4242/#"));
+  assertThrows(() => new WebSocketStream("ws://localhost:4242/#foo"));
+});
+
+Deno.test("duplicate protocols", () => {
+  assertThrows(() =>
+    new WebSocketStream("ws://localhost:4242", {
+      protocols: ["foo", "foo"],
+    })
+  );
+});
+
+Deno.test("invalid server", async () => {
+  const ws = new WebSocketStream("ws://localhost:2121");
+  await Promise.all([
+    assertThrowsAsync(() => ws.connection),
+    assertThrowsAsync(() => ws.closed),
+  ]);
+});
+
+Deno.test("connect & close", async () => {
+  const ws = new WebSocketStream("ws://localhost:4242");
+  await ws.connection;
+  ws.close();
+  await ws.closed;
+});
+
+Deno.test("connect & close custom valid code", async () => {
+  const ws = new WebSocketStream("ws://localhost:4242");
+  await ws.connection;
+  ws.close({ code: 1000 });
+  await ws.closed;
+});
+
+Deno.test("connect & close custom invalid code", async () => {
+  const ws = new WebSocketStream("ws://localhost:4242");
+  await ws.connection;
+  assertThrows(() => ws.close({ code: 1001 }));
+  ws.close();
+  await ws.closed;
+});
+
+Deno.test("connect & close custom valid reason", async () => {
+  const ws = new WebSocketStream("ws://localhost:4242");
+  await ws.connection;
+  ws.close({ code: 1000, reason: "foo" });
+  await ws.closed;
+});
+
+Deno.test("connect & close custom invalid reason", async () => {
+  const ws = new WebSocketStream("ws://localhost:4242");
+  await ws.connection;
+  assertThrows(() => ws.close({ code: 1000, reason: "".padEnd(124, "o") }));
+  ws.close();
+  await ws.closed;
+});
+
+Deno.test("echo string", async () => {
+  const ws = new WebSocketStream("ws://localhost:4242");
+  const { readable, writable } = await ws.connection;
+  await writable.getWriter().write("foo");
+  const res = await readable.getReader().read();
```

Comment:
The "leaking" promise is created here.
`read()` returns a promise which never resolves and doesn't reject either.

Reply:
It's actually fairly easy to replicate in the CLI as well:


```js
D:\deno>target\debug\deno.exe
Deno 1.9.2
exit using ctrl+d or close()
>   const ws = new WebSocketStream("ws://localhost:4242");
undefined
>   const { readable, writable } = await ws.connection;
undefined
>   const uint = new Uint8Array([102, 111, 111]);
undefined
>   await writable.getWriter().write(uint);
undefined
>   const res = await readable.getReader().read();
... hangs ...
```